### PR TITLE
Add new BBE for Binding pattern in assignment

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose
$subject

### CODE REVIEW

File: examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+function assign(IntPair ip) {
+    // The variable `ip` of type `IntPair` is destructured and
+    // assigned to the two module-level variables `x` and `y`.
+    [x, y] = ip;
+
+    io:println(x);
+    io:println(y);
+
+    swap();
```
Comment (KRVPerera) [inline examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal:16]: This is a shortcut. So let's not use a separate function here.

File: examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+int x = 0;
+int y = 1;
+
+type IntPair [int, int];
+
+function assign(IntPair ip) {
+    // The variable `ip` of type `IntPair` is destructured and
+    // assigned to the two module-level variables `x` and `y`.
+    [x, y] = ip;
```
Comment (KRVPerera) [inline examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal:11]: Shall we use some local variables here like 
```suggestion
    [p, q] = ip;
```

File: examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+function swap() {
+    // The swapping is done between `x` and `y` just by doing
+    // tuple destructuring without using a temporary variable
+    [x, y] = [y, x];
+}
+
+public function main() {
+    assign([x, y]);
```
Comment (KRVPerera) [inline examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal:29]: Lets remove global variables
```suggestion
    assign([4, 5]);
```

File: examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal (hunk 2 above)
Comment (KRVPerera) [inline examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal:11]: This is also a cool trick
```suggestion
    [_, z] = ip;
```

File: examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal (hunk 1 above)
Comment (KRVPerera) [inline examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal:16]:
```suggestion
    [p, q] = [q, p];
```

File: examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+}
```
Comment (KRVPerera) [inline examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal:30]: This feature is coming from modern Java Script, I think. 

Some more cool ideas
```Ballerina
    int a;
    int b;
    int c;
    [a, b, _, c] = [...[1, 2], ...[3, 4]];
```
`1
2
4`

Comment (KRVPerera) [reply examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal:30]:
```Ballerina
    int a;
    int b;
    int d;
    int e;
    int[] rest;
    [a, b, _, d, e, ...rest] = [...[1, 2], ...[3, 4], ...[5, 45, 345]];
    io:println(a);
    io:println(b);
    io:println(d);
    io:println(e);
    io:println(rest);
```

`1
2
4
5
[45,345]`

Comment (KRVPerera) [reply examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal:30]:
```Ballerina
    int[] a1 = [1, 2];
    int[] b1 = [3, 4];
    [a1, b1] = [b1, a1];

    io:println(a1, b1);
```

**Review by KRVPerera — COMMENTED** [review]
I think this BBE only demonstrate list restructuring. 
We need this kind of example too.

```Ballerina

type FullName record {
    string firstName;
    string middleName;
    string lastName;
};

public function main() {
  FullName fullNameOfStudent = {
          firstName: "Jane",
          middleName: "unknown",
          lastName: "Doe"
      };

    string firstName;
    string lastName;

    {
        firstName,
        lastName
    } = fullNameOfStudent;

    io:println(firstName);
    io:println(lastName);
}
```

**Comment by KavinduZoysa** [comment]
> I think this BBE only demonstrate list destructuring. We need this kind of example too.
> 
> ```ballerina
> type FullName record {
>     string firstName;
>     string middleName;
>     string lastName;
> };
> 
> public function main() {
>   FullName fullNameOfStudent = {
>           firstName: "Jane",
>           middleName: "unknown",
>           lastName: "Doe"
>       };
> 
>     string firstName;
>     string lastName;
> 
>     {
>         firstName,
>         lastName
>     } = fullNameOfStudent;
> 
>     io:println(firstName);
>     io:println(lastName);
> }
> ```

Added

File: examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal (hunk 4 above)
Comment (KavinduZoysa) [reply examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal:64]: Added few more cases

File: examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal (hunk 1 above)
Comment (KavinduZoysa) [reply examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal]: Changed

File: examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal (hunk 2 above)
Comment (KavinduZoysa) [reply examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal]: Changed

File: examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal (hunk 3 above)
Comment (KavinduZoysa) [reply examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal]: Changed

File: examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal (hunk 2 above)
Comment (KavinduZoysa) [reply examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal]: Changed

File: examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal (hunk 1 above)
Comment (KavinduZoysa) [reply examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal]: Changed

**Review by KRVPerera — COMMENTED** [review]
LGTM

File: examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal [hunk 5]
```diff
@@ -0,0 +1,64 @@
+import ballerina/io;
+
+type IntPair [int, int];
+
+function assignToListBindingPattern(IntPair ip) {
+    int p;
+    int q;
+    // The variable `ip` of type `IntPair` is destructured and
```
Comment (praneesha) [inline examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal:8]:
```suggestion
    // The `ip` variable of the IntPair `type` is destructured and
```

File: examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    // assigned to the two variables `p` and `q`.
```
Comment (praneesha) [inline examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal:9]:
```suggestion
    // assigned to the two `p` and `q` variables.
```

File: examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    [p, q] = ip;
+
+    io:println(p);
+    io:println(q);
+
+    // The swapping is done between `p` and `q` just by doing
```
Comment (praneesha) [inline examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal:15]:
```suggestion
    // The swapping is done between `p` and `q` by doing
```

File: examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    // tuple destructuring without using a temporary variable
```
Comment (praneesha) [inline examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal:16]:
```suggestion
    // tuple destructuring without using a temporary variable.
```

File: examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+    io:println(r);
+
+    int a;
+    int b;
+    int d;
+    int e;
+    int[] rest;
+
+    // Destructure a list which includes rest value.
```
Comment (praneesha) [inline examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal:34]:
```suggestion
    // Destructure a list that includes a `rest` value.
```

File: examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    string firstName;
+    string middleName;
+    string lastName;
+};
+
+function assignToMappingBindingPattern(FullName fullName) {
+    string firstName;
+    string lastName;
+
+    // The variable `fullName` of type `FullName` is destructured and
```
Comment (praneesha) [inline examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal:53]:
```suggestion
    // The `fullName` variable of  the `FullName` type is destructured and
```

File: examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    // assigned to the two variables `firstName` and `lastName`.
```
Comment (praneesha) [inline examples/binding-patterns-in-assignment/binding_patterns_in_assignment.bal:54]:
```suggestion
    // assigned to the two `firstName` and `lastName` variables.
```

**Review by praneesha — APPROVED** [review]

**Review by chiranSachintha — APPROVED** [review]

**Pull request merged**